File: ProiectCS/ProiectCS/repository/TicketDBRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using NET.model;
using Npgsql;


namespace NET.repository
{
    public class TicketDBRepository : TicketRepository
    {
        private static readonly ILog log = LogManager.GetLogger("TicketDBRepository");
        
        private IDictionary<string, string> props;

        public TicketDBRepository(IDictionary<string, string> props)
        {
            log.Info("Creating TiketDBRepository");
            this.props = props;
        }

        public void Save(Ticket elem)
        {
            log.InfoFormat("saving ticket {0}", elem);
            using (var connection = new NpgsqlConnection(props["ConnectionString"]))
            {
                connection.Open();
                using (var saveStmt = new NpgsqlCommand("insert into Tickets(customerName,touristName,customerAddress,seats,FlightID)" +
                                                        "values (@cN,@tN,@cA,@seats,@FlightID)", connection))
                {
                    saveStmt.Parameters.AddWithValue("@cN", elem.CustomerName);
                    saveStmt.Parameters.AddWithValue("@tN", elem.TouristName);
                    saveStmt.Parameters.AddWithValue("@cA", elem.CustomerAddress);
                    saveStmt.Parameters.AddWithValue("@seats", elem.Seats);
                    saveStmt.Parameters.AddWithValue("@FlightID", elem.FlightForTicket.ID);
                    var result = saveStmt.ExecuteNonQuery();
                    if (result == 0)
                    {
                        log.Error("No ticket added !");
                        //throw new Exception("No ticket added !");
                    }
                }
            }
            log.InfoFormat("ticket has been added {0}", elem);
        }

        public void Delete(Ticket elem)
        {
            log.InfoFormat("deleting ticket {0}", elem);
            using (var connection =  new NpgsqlConnection(props["ConnectionString"]))
            {
                connection.Open();
                using (var deleteStmt = new NpgsqlCommand("delete from Tickets where id = @id", connection))
                {
                    deleteStmt.Parameters.AddWithValue("@id", elem.ID);
                    var result = deleteStmt.ExecuteNonQuery();
                    if (result == 0)
                    {
                        log.Error("No ticket deleted");
                        //throw new Exception("No ticket deleted !");
                    }
                }
            }
            log.InfoFormat("ticket has been deleted {0}", elem);
        }

        public void Update(Ticket elem, int ID)
        {
            log.InfoFormat("updating ticket {0} with ID {1}", elem, ID);
            using (var connection = new NpgsqlConnection(props["ConnectionString"]))
            {
                connection.Open();
                using (var updateStmt = new NpgsqlCommand("update Tickets set customerName = @cN , touristName = @tN ," +
                                                          "customerAddress = @cA,seats = @seats,FlightID = @FlightID) where id = @id", connection))
                {

                    updateStmt.Parameters.AddWithValue("@cN", elem.CustomerName);
                    updateStmt.Parameters.AddWithValue("@tN", elem.TouristName);
                    updateStmt.Parameters.AddWithValue("@cA", elem.CustomerAddress);
                    updateStmt.Parameters.AddWithValue("@seats", elem.Seats);
                    updateStmt.Parameters.AddWithValue("@FlightID", elem.FlightForTicket.ID);
                    updateStmt.Parameters.AddWithValue("@id", ID);
                    var result = updateStmt.ExecuteNonQuery();
                    if (result == 0)
                    {
                        log.Error("No ticket updated!");
                        //throw new Exception("No ticket updated!");
                    }
                }
            }
            log.InfoFormat("ticket has been updated {0} with ID {1}", elem, ID);
        }

        public Ticket FindById(int ID)
        {
            log.InfoFormat("finding ticket with ID {0}", ID);
            using (var connection = new NpgsqlConnection(props["ConnectionString"]))
            {
                connection.Open();
                using (var findStmt = new NpgsqlCommand("select * from Tickets  T inner join Flights  F on T.Flightid = F.id where T.id = @id", connection))
                {
                    findStmt.Parameters.AddWithValue("@id", ID);
                    using (var dataR = findStmt.ExecuteReader())
                    {
                        if (dataR.Read())
                        {
                            string customerName = dataR.GetString(1);
                            string touristName = dataR.GetString(2);
                            string customerAddress = dataR.GetString(3);
                            int seats = dataR.GetInt32(4);
                            int idTicket = dataR.GetInt32(6);
                            string destination = dataR.GetString(7);
                            DateTime departure = dataR.GetDateTime(8);
                            string airport = dataR.GetString(9);
                            int availableSeats = dataR.GetInt32(10);

                            Flight flight = new Flight(idTicket, destination, departure, airport, availableSeats);
                            Ticket ticket = new Ticket(ID, customerName, touristName, customerAddress, seats, flight);
                            log.InfoFormat("ticket with id {0} has been found {1}", ID, ticket);
                            return ticket;
                        }
                    }
                }
                log.InfoFormat("no ticket with id {0}", ID);
            }

            return null;
        }

        public IEnumerable<Ticket> FindAll()
        {
            log.Info("finding all tickets");
            IList<Ticket> tickets = new List<Ticket>();
            using (var connection = new NpgsqlConnection(props["ConnectionString"]))
            {
                connection.Open();
                using (var findStmt = new NpgsqlCommand("select * from Tickets AS T inner join Flights AS F on T.FlightID = F.id", connection))
                {
                    using (var dataR = findStmt.ExecuteReader())
                    {
                        while (dataR.Read())
                        {
                            int ID = dataR.GetInt32(0);
                            string customerName = dataR.GetString(1);
                            string touristName = dataR.GetString(2);
                            string customerAddress = dataR.GetString(3);
                            int seats = dataR.GetInt32(4);
                            int idTicket = dataR.GetInt32(6);
                            string destination = dataR.GetString(7);
                            DateTime departure = dataR.GetDateTime(8);
                            string airport = dataR.GetString(9);
                            int availableSeats = dataR.GetInt32(10);

                            Flight flight = new Flight(idTicket, destination, departure, airport, availableSeats);
                            Ticket ticket = new Ticket(ID, customerName, touristName, customerAddress, seats, flight);
                            tickets.Add(ticket);
                        }
                    }
                    log.Info("return all found tickets");
                    return tickets;
                }
            }
        }
    }
}

File: Java/src/main/java/repository/TicketDBRepository.java
package repository;

import model.Flight;
import model.Ticket;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;


public class TicketDBRepository implements TicketRepository{

    private final static Logger logger = LogManager.getLogger(TicketDBRepository.class);
    private JdbcUtils dbUtils;
    private FlightRepository flightRepository;

    public TicketDBRepository(FlightRepository flightRepository){
        logger.info("creating TicketDBRepository");
        this.dbUtils = new JdbcUtils();
        this.flightRepository = flightRepository;
    }

    @Override
    public void add(Ticket elem) {
        logger.traceEntry("save ticket {} ", elem);
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement insertStmt = connection
                    .prepareStatement("insert into Tickets(customerName,touristName,customerAddress,seats,FlightID) " +
                            "values (?,?,?,?,?)")) {
            insertStmt.setString(1, elem.getCustomerName());
            insertStmt.setString(2,elem.getTouristName());
            insertStmt.setString(3,elem.getCustomerAddress());
            insertStmt.setInt(4,elem.getSeats());
            insertStmt.setInt(5,elem.getFlight().getID());
            insertStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Ticket elem) {
        logger.traceEntry("delete ticket {} ", elem);
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement deleteStmt = connection
                    .prepareStatement("delete from Tickets where id = ?")) {
            deleteStmt.setInt(1,elem.getID());
            deleteStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Ticket elem, Integer id) {
        logger.traceEntry("update ticket {} with id {}", elem, id);
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement updateStmt = connection
                    .prepareStatement("update Tickets set customerName = ? ,touristName = ? " +
                            ",customerAddress = ? ,seats = ?, FlightID = ? where id = ?")) {
            updateStmt.setString(1,elem.getCustomerName());
            updateStmt.setString(2,elem.getTouristName());
            updateStmt.setString(3,elem.getCustomerAddress());
            updateStmt.setInt(4,elem.getSeats());
            updateStmt.setInt(5,elem.getFlight().getID());
            updateStmt.setInt(6,id);
            updateStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit();
    }

    @Override
    public Ticket findByID(Integer id) {
        logger.traceEntry("finding ticket with id {}", id);
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement findStmt = connection
                    .prepareStatement("select * from Tickets T inner join Flights F " +
                            "on T.FlightID = F.id where id = ?")) {
            findStmt.setInt(1,id);
            ResultSet resultSet = findStmt.executeQuery();
            if(resultSet.next()){
                Ticket ticket = buildTicket(resultSet);;
                return ticket;
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit("No ticket found with id {}", id);
        return null;
    }

    @Override
    public Iterable<Ticket> findAll() {
        logger.traceEntry("find all tickets");
        List<Ticket> ticketList = new ArrayList<>();
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement selectStmt = connection
                    .prepareStatement("select * from Tickets T inner join Flights F on T.FlightID = F.id")) {
            ResultSet resultSet = selectStmt.executeQuery();
            while (resultSet.next()){
                Ticket ticket = buildTicket(resultSet);
                ticketList.add(ticket);
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
        return ticketList;
    }

    @Override
    public Collection<Ticket> getAll() {
        logger.traceEntry("get all tickets");
        List<Ticket> ticketList = new ArrayList<>();
        try(Connection connection = dbUtils.getConnection();
            PreparedStatement selectStmt = connection
                    .prepareStatement("select * from Tickets T inner join Flights F on T.FlightID = F.id")) {
            ResultSet resultSet = selectStmt.executeQuery();
            while (resultSet.next()){
                Ticket ticket = buildTicket(resultSet);
                ticketList.add(ticket);
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit();
        return ticketList;
    }

    private Ticket buildTicket(ResultSet resultSet) throws SQLException {
        Integer id = resultSet.getInt("id");
        String customerName = resultSet.getString("customerName");
        String touristName = resultSet.getString("touristName");
        String customerAddress = resultSet.getString("customerAddress");
        Integer seats = resultSet.getInt("seats");
        Integer flightID = resultSet.getInt("FlightID");

        Flight flight = flightRepository.findByID(flightID);
        Ticket ticket = new Ticket(id,customerName,touristName,customerAddress,seats,flight);
        return ticket;
    }
}
